site/src/components/WorkspaceStatusBadge/WorkspaceStatusBadge.tsx

```diff
@@ -0,0 +1,170 @@
+import CircularProgress from "@material-ui/core/CircularProgress"
+import { makeStyles, Theme, useTheme } from "@material-ui/core/styles"
+import ErrorIcon from "@material-ui/icons/ErrorOutline"
+import PlayIcon from "@material-ui/icons/PlayArrowOutlined"
+import StopIcon from "@material-ui/icons/StopOutlined"
+import { WorkspaceBuild } from "api/typesGenerated"
+import React from "react"
+import { MONOSPACE_FONT_FAMILY } from "theme/constants"
+import { combineClasses } from "util/combineClasses"
+import { getWorkspaceStatus } from "util/workspace"
+
+const StatusLanguage = {
+  loading: "Loading",
+  started: "Running",
+  starting: "Starting",
+  stopping: "Stopping",
+  stopped: "Stopped",
+  deleting: "Deleting",
+  deleted: "Deleted",
+  canceling: "Canceling action",
+  canceled: "Canceled action",
+  failed: "Failed",
+  queued: "Queued",
+}
+
+const LoadingIcon: React.FC = () => {
+  return <CircularProgress size={10} style={{ color: "#FFF" }} />
+}
+
+export const getStatus = (
+  theme: Theme,
+  build: WorkspaceBuild,
+): {
+  borderColor: string
+  backgroundColor: string
+  text: string
+  icon: React.ReactNode
+} => {
+  const status = getWorkspaceStatus(build)
+  switch (status) {
```

Comment:
I wonder if we can key off the `WorkspaceStateEnum` we created the other day: https://github.com/coder/coder/blob/main/site/src/components/WorkspaceActions/constants.ts#L2
Reasoning: that way we can have 1 definitive source of truth for workspace state. For example, I don't see `loading` in this switch (maybe it's not needed?), although it is a state that can be returned from the API.

Reply:
I don't think so. It is because the status is derivative and sometimes it will use the status + transition combinations to get the right result to display but I agree it is bad having this logic spread around. IMO, the best solution is to have these computed in the BE and returned by the API. Maybe related to this https://github.com/coder/coder/issues/2772

Reply:
Agree. Looks like #1720 might address this as well. I'll see if we can close one of them out.